### PR TITLE
[Bugfix] Fix deepseek_v2 missing last layer aux hidden state extraction

### LINKED ISSUE #48124: [Bug]: The deepseek_v2.py function misses the output activation values ​​of the last layer when extracting hidden states.
There will be no operations to extract hidden state before and after the last layer. <code>model_executor/models/deepseek_v2.py 

### 🐛 Describe the bug

The forward of <code> model_executor/models/deepseek_v2.py </code> code as below 
It can be observed that if there are 60 hidden layers, there will be no operation to extract hidden state after the 60th layer forward, which will affect the training of models such as dflash (the last layer activation value is needed to calculate the loss).
```python
        aux_hidden_states = []
        for idx, layer in enumerate(
            islice(self.layers, self.start_layer, self.end_layer),
            start=self.start_layer,
        ):
            # all gather if we need to use the whole states
            if (
                hidden_states.shape[0] != positions.shape[0]
                and not layer.use_sequence_parallel_moe
            ):
                combined_states = torch.cat([hidden_states, residual], dim=-1)
                combined_states = tensor_model_parallel_all_gather(combined_states, 0)
                combined_states = combined_states[: positions.shape[0]]
                hidden_states, residual = combined_states.split(
                    [self.hidden_size, self.hidden_size], dim=-1
                )
                # fused_add_rms_norm requires a contiguous residual
                residual = residual.contiguous()
            if idx in self.aux_hidden_state_layers:
                aux_hidden_state = hidden_states + residual
                if aux_hidden_state.shape[0] != positions.shape[0]:
                    aux_hidden_state = tensor_model_parallel_all_gather(
                        aux_hidden_state, 0
                    )
                    aux_hidden_state = aux_hidden_state[: positions.shape[0]]
                aux_hidden_states.append(aux_hidden_state)
            hidden_states, residual = layer(
                positions, hidden_states, residual, llama_4_scaling
            )
```
[https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/deepseek_v2.py](url)
fix eg:
```python
            if idx + 1 in self.aux_hidden_state_layers:
                aux_hidden_states.append(hidden_states + residual)
```

### PR DESCRIPTION
**Pull request by factnn**
Move the aux_hidden_state capture in the forward loop from before layer()
to after it, and use idx + 1 so layer outputs are captured under 1-based
layer indices. Without this, the last layer's output is never appended to
aux_hidden_states, which breaks dflash training that needs the final layer
activation. Matches the pattern in deepseek_v4/nvidia/model.py.

Fixes #48124

### CODE REVIEW

**Comment by mgoin** [comment]
Can you explain why https://github.com/vllm-project/vllm/pull/46973 is not sufficient? Once I landed my PR, I was able to train DSpark for GLM 5.2 fine with Speculators

**Review by ErenAta16 — COMMENTED** [review]
Traced through the indexing carefully before commenting, since moving a check across a function call and flipping its comparison by one is easy to get subtly wrong in either direction.

For any layer index reachable purely inside the loop body, checking `idx in aux_hidden_state_layers` before `layer(idx)` runs and checking `idx + 1 in aux_hidden_state_layers` after it runs are algebraically the same condition (both reduce to "capture layer k's output when k+1 is in the set"), so by itself the reordering doesn't change what gets captured for those indices.

Where it does matter, and where I think the real fix is: `idx` only ever reaches `num_layers - 1` inside the loop, never `num_layers`. So under the old code, a target value of `num_layers` (the loop bound itself, used to mean "the very last layer") could only ever be satisfied by the separate post-loop special case (`if self.end_layer in self.aux_hidden_state_layers`), never by the in-loop check. Under the new code, that same target is now *also* satisfied in-loop, at the last iteration (`idx + 1 == num_layers` when `idx == num_layers - 1`). That overlap is exactly why the post-loop branch had to change from an unconditional `append` to `aux_hidden_states[-1] = ... if aux_hidden_states else append(...)` -- without that guard, a config that wants the final layer's output via this convention would get it appended twice.

I checked this concretely: simulating the literal one-liner suggested in #48124 (`if idx + 1 in self.aux_hidden_state_layers: aux_hidden_states.append(...)`, with the post-loop append left untouched) against `aux_hidden_state_layers = {num_layers}` on an 8-layer stand-in reproduces a duplicate final entry (`['L7', 'L7']` instead of `['L7']`). This PR's version returns the correct single entry, so the overwrite guard is doing real work here, not just defensive code.

One thing I couldn't fully resolve and would like to ask about: in my trace, the pre-existing post-loop branch already looks like it should correctly capture `num_layers` as a standalone target on a single rank (no pipeline parallelism), which would mean the original single-GPU case wasn't actually broken for that specific config, before this PR, in the way #48124's summary suggests ("no operation to extract hidden state after the 60th layer forward"). Was the repro that motivated this running under pipeline parallelism, or hitting some other code path where `self.end_layer` doesn't line up with the global layer count, or a case with an early return before reaching the post-loop branch? Not blocking, just want to understand the exact failure mode so it's clear this covers it.

Separately: confirmed the `idx + 1 in self.aux_hidden_state_layers` check-after-layer() shape is already the established pattern in `vllm/models/deepseek_v4/nvidia/model.py`, so this brings deepseek_v2 in line with existing precedent either way.

**Comment by ErenAta16** [comment]
@mgoin I think I can answer this, since I went through the exact same "is this actually still broken" question while reviewing.

#46973 is where the post-loop block in the current code came from:

```python
if self.end_layer in self.aux_hidden_state_layers:
    aux_hidden_states.append(hidden_states + residual)
```

That block alone already correctly captures a target value equal to `num_layers` (the "final layer" case #46973's description describes) as a standalone target on a single rank -- I checked this by simulating the exact pre-#48152 loop logic, and a config with `aux_hidden_state_layers = {num_layers}` returns the correct single final-layer entry with today's code, no `#48152` needed. So `#48124`'s framing ("no operation to extract hidden state after the last layer") isn't accurate anymore against current main; #46973 already covers that specific case.

Where I do think this PR still does real work: once you move the in-loop check to `idx + 1 in aux_hidden_state_layers` (checked after `layer(idx)` runs, as `#48124`'s suggested one-liner does), `idx` can reach `num_layers - 1` on the last loop iteration, and `idx + 1 == num_layers` then satisfies the *same* condition `#46973`'s post-loop block also checks. Applying `#48124`'s one-liner in isolation, without also touching the post-loop append, would append the final layer's state twice for that config -- I confirmed this concretely by simulating both versions side by side. This PR's added `aux_hidden_states[-1] = ... if aux_hidden_states else append(...)` guard is what prevents that double-append.

So as it stands, this PR mostly guards against a duplicate-entry bug that would only surface if someone applied the naive fix from #48124 without this PR's extra guard -- it's not clear to me it's fixing a bug that's actually reachable on current main (post-#46973) as written. @anxkhn might be worth confirming whether the original dflash repro was run against a version of main that predates #46973, since that would explain the discrepancy.

**Comment by factnn** [comment]
Thanks for the review and the detailed explanation @mgoin @ErenAta16 — I missed that #46973 already handles this case. Closing.
